### PR TITLE
[CONTRIB] Allow customized initializer in PopenPool

### PR DESCRIPTION
**Pull request by shingjan**
This PR intends to add `initializer` as well as `initargs` to `popenpool` api.

@tqchen @vinx13 @junrushao1994

### CODE REVIEW

File: python/tvm/contrib/popen_pool.py
Comment (vinx13) [inline python/tvm/contrib/popen_pool.py:89]: please add docs for the args

File: python/tvm/contrib/popen_pool.py
Comment (vinx13) [inline python/tvm/contrib/popen_pool.py:135]: remove this

**Comment by junrushao1994** [comment]
Thanks @shingjan for the contribution! This work definitely makes Popen more complete, in terms of customizable initialization of workers :-)

BTW, please document the usage of this argument, so that users could understand what is going on under the hood

File: python/tvm/contrib/popen_pool.py
Comment (vinx13) [inline python/tvm/contrib/popen_pool.py:291]: please add docs for the args

File: python/tvm/testing/popen_pool.py
Comment (vinx13) [inline python/tvm/testing/popen_pool.py:20]: consider using meaningful variable names

**Review by vinx13 — CHANGES_REQUESTED** [review]

**Comment by vinx13** [comment]
Thanks @shingjan , this will unblock refactor of using PopenPool in autotvm

File: tests/python/contrib/test_popen_pool.py
Comment (vinx13) [inline tests/python/contrib/test_popen_pool.py:71]: `pool` is not used

File: python/tvm/contrib/popen_pool.py
Comment (junrushao1994) [inline python/tvm/contrib/popen_pool.py:166]: I am not sure about the logic here. Looks like the workflow is:

* `PopenWorker.send` is the entry point, which calls `PopenWorker._start` if the worker is not available
* with this logic, the `PopenWorker._start` method will call back to `PopenWorker.send` to send the initializer to the worker process.

I didn't see any problem with the workflow itself, but the cyclic call is a bit indirect. Would you mind if we move these 4 lines into the `if self._proc is None` check?

**Comment by junrushao1994** [comment]
BTW, if we have a file containing user-registered functions, can we send this to the worker process so that the worker is aware of these registration? I suppose it is as simple as sending an "import XXX"

**Comment by vinx13** [comment]
> BTW, if we have a file containing user-registered functions, can we send this to the worker process so that the worker is aware of these registration? I suppose it is as simple as sending an "import XXX"

In this case the registry map need to be sent. Sending an 'import' may also work, it requires the user file that calls register_func be available in PYTHONPATH

**Comment by junrushao1994** [comment]
@vinx13 this snippet does work, but i am not sure if it is the best way to do:

```python
import importlib.util
spec = importlib.util.spec_from_file_location("mymodule", "/home/jrshao/Projects/tvm-tensorir-dev/tests/python/tir/util.py")
foo = importlib.util.module_from_spec(spec)
spec.loader.exec_module(foo)
```

**Comment by junrushao1994** [comment]
@shingjan please address the comments and add a testcase for registering extra methods with `tvm._ffi.register_func` and `tvm._ffi.get_global_func`. Let's get it merged ASAP :-)

File: python/tvm/contrib/popen_pool.py
Comment (shingjan) [reply python/tvm/contrib/popen_pool.py:166]: done

File: tests/python/contrib/test_popen_pool.py
Comment (shingjan) [reply tests/python/contrib/test_popen_pool.py:71]: removed

File: python/tvm/testing/popen_pool.py
Comment (shingjan) [reply python/tvm/testing/popen_pool.py:20]: done

File: python/tvm/contrib/popen_pool.py
Comment (shingjan) [reply python/tvm/contrib/popen_pool.py:135]: done

File: python/tvm/contrib/popen_pool.py
Comment (shingjan) [reply python/tvm/contrib/popen_pool.py:89]: done

File: python/tvm/contrib/popen_pool.py
Comment (shingjan) [reply python/tvm/contrib/popen_pool.py:291]: done

**Comment by shingjan** [comment]
@junrushao1994 I added three new test cases for `ffi`, including calling python function via ffi, calling cpp function via ffi and calling python function from cpp function via ffi in `popenworker`.

File: tutorials/autotvm/tune_conv2d_cuda.py
Comment (junrushao1994) [inline tutorials/autotvm/tune_conv2d_cuda.py:207]: probably we don't need this :-)

**Comment by junrushao1994** [comment]
Overall it looks pretty good now :-) Please fix the lint and let's get it merged ASAP

File: tutorials/autotvm/tune_conv2d_cuda.py
Comment (shingjan) [reply tutorials/autotvm/tune_conv2d_cuda.py]: it is removed. I was changing this for benchmarking. Slipping through my finger lol

**Review by junrushao1994 — APPROVED** [review]
LGTM. Thanks for the contribution @shingjan!

**Comment by junrushao1994** [comment]
Oh looks like I need @vinx13's approval to merge this PR, but he is now on vacation 😄

**Comment by shingjan** [comment]
I think Github used to let contributors revoke `request for changes`. But maybe that is not the case for public repository like TVM.

**Comment by tqchen** [comment]
you can dismiss the reviews as long as the comments are addressed

File: python/tvm/contrib/popen_pool.py
Comment (vinx13) [inline python/tvm/contrib/popen_pool.py:272]: Keep this line

**Review by vinx13 — APPROVED** [review]
Overall looks good except some minor changes

File: python/tvm/contrib/popen_pool.py
Comment (shingjan) [reply python/tvm/contrib/popen_pool.py:272]: Thanks @junrushao1994 for adding this line and @vinx13 for the comments!

**Pull request merged**
